### PR TITLE
[xh] Add oracle source

### PR DESCRIPTION
**Pull request by matrixstone**
# Summary
Add oracle source.

Still testing with bookmark function.

# Tests
Tested with Mage UI

## Show configuration
oracle_show_config

## Test connection success

oracle_test_connection_success

## View tables success
oracle_view_streams

## Show table schemas
oracle_view_table_schema

## Load sample data
oracle_load_sample_data

## Pipeline run completes

Run complete

## Data exported to S3
oracle_output_to_s3

## Data content exported to S3
oracle_exported_data_in_csv

cc:
@wangxiaoyou1993 @dy46

### CODE REVIEW

File: mage_integrations/mage_integrations/sources/oracledb/__init__.py [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+            elif c_name in pk_constraints.get(table_schema, {}).get(table_name, []):
+                mdata = metadata.write(mdata, ('properties', c_name), 'inclusion', 'automatic')
+                mdata = metadata.write(mdata, ('properties', c_name), 'selected-by-default', True)
+            else:
+                mdata = metadata.write(mdata, ('properties', c_name), 'inclusion', 'available')
+                mdata = metadata.write(mdata, ('properties', c_name), 'selected-by-default', True)
+
+        return mdata
+
+    def schema_for_column(self, c, pks_for_table):
```
Comment (wangxiaoyou1993) [inline mage_integrations/mage_integrations/sources/oracledb/__init__.py:119]: I think your implementation can be simplified a lot if you inherit the sql source.
You can just override the build_discover_query method like this https://github.com/mage-ai/mage-ai/blob/master/mage_integrations/mage_integrations/sources/mssql/__init__.py#L29-L54

**Comment by wangxiaoyou1993** [comment]
Could you set up pre-commit following the instructions in this doc: https://github.com/mage-ai/mage-ai/blob/master/README_dev.md? This helps check your code style before commit.

File: mage_integrations/mage_integrations/sources/oracledb/__init__.py (hunk 1 above)
Comment (matrixstone) [reply mage_integrations/mage_integrations/sources/oracledb/__init__.py]: Thank you for the suggestion Xiaoyou! It is a much concise solution. Updated with overwriting of `build_discover_query` function.

**Comment by matrixstone** [comment]
Thank you for the review! Updated the PR following comment and setted up git precommit. PTAL. @wangxiaoyou1993

File: mage_integrations/mage_integrations/sources/oracledb/README.md [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+## Config
+
+You must enter the following credentials when configuring this source:
+
+| Key | Description | Sample value
+| --- | --- | --- |
+| `host` | OracleDB host. | `oracledb.example.com` |
+| `port` | OracleDB exported host. | `1521` (default value) |
+| `service` | 
+Listener service runs on the database server listing for client connection. | `xepdb1` |
```
Comment (wangxiaoyou1993) [inline mage_integrations/mage_integrations/sources/oracledb/README.md:12]: nit: you don't need to change the line here

File: mage_integrations/mage_integrations/sources/oracledb/__init__.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            table_names = ', '.join([f"'{n}'" for n in streams])
+            query = query.format(where_table_clause=f"\nAND user_tab.TABLE_NAME IN ({table_names})")
+        else:
+            query = query.format(where_table_clause='')
+
+        return query
+
+    def build_connection(self) -> OracleDBConnection:
+        return OracleDBConnection(
+            self.host, self.password, self.user, self.port, self.service)
```
Comment (wangxiaoyou1993) [inline mage_integrations/mage_integrations/sources/oracledb/__init__.py:76]: for keyword arguments, please use `port=self.port, service=self.service`

**Review by wangxiaoyou1993 — APPROVED** [review]

File: mage_integrations/mage_integrations/sources/oracledb/README.md (hunk 2 above)
Comment (matrixstone) [reply mage_integrations/mage_integrations/sources/oracledb/README.md]: Updated

File: mage_integrations/mage_integrations/sources/oracledb/__init__.py (hunk 3 above)
Comment (matrixstone) [reply mage_integrations/mage_integrations/sources/oracledb/__init__.py]: Updated

**Pull request merged**